core/src/main/java/io/grpc/InternalInstrumented.java

```diff
@@ -0,0 +1,30 @@
+/*
+ * Copyright 2017, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc;
+
+import java.util.concurrent.Future;
+
+/**
+ * This is an gRPC internal interface. Do not use this.
+ *
+ * <p>An interface for types that <b>may</b> support instrumentation. If the actual type does not
+ * support instrumentation, then the future will return a {@code null}.
+ */
+@Internal
+public interface InternalInstrumented<T> extends InternalWithLogId {
```

Comment:
FYI, nothing in `^io.grpc$` can reference this class, since it would cause dependency cycles.  Can you give a sample usage of this inside `io.grpc` ?

Reply:
Moved the classes around to achieve proper visibility. This will be a sketch of the final result:

```
"core": package io.grpc
===============
// pkg-private so they are hidden from users

class StatsServerImpl {
  private static final StatsServerImpl INSTANCE = new StatsServerImpl();

  private StatsServerImpl() { /* do not instantiate */ }

  // all methods package-private
  static StatsServerImpl getInstance() { return INSTANCE; }
  void registerServer(Instrumented<ServerStats> server) { /*TODO*/ }
  void registerChannel(Instrumented<ChannelStats> server) { /*TODO*/ }
  void registerTransport(Instrumented<TransportStats> server) { /*TODO*/ }
}

class LogId {
  public long getId() { /* TODO */ }
}

interface WithLogId {
  LogId getId();
}

interface Instrumented<T> {
  T getStat();
}
class ServerStats { /*TODO*/ }
class ChannelStats { /*TODO*/ }
class TransportStats { /*TODO*/ }

"core_internal_accessor": package io.grpc
=====================================

// hole punch so that io.grpc.internal can implement LogId and WithLogId

public class InternalLogId extends LogId { }
public interface InternalWithLogId extends WithLogId {
  @Override
  public InternalLogId getId();
}

// hole punch so that netty and others can return stats

public interface InternalInstrumented<T> extends Instrumented<T>, InternalWithLogId { /*TODO*/ }
public class InternalServerStats extends ServerStats { /*TODO*/ }
public class InternalChannelStats extends ChannelStats { /*TODO*/ }
public class InternalTransportStats extends TransportStats { /*TODO*/ }

"netty": package io.grpc.netty
=====================

class NettyServer implements InternalInstrumented<ServerStats> { /*TODO*/ }
class NettyClientTransport implements InternalInstrumented<InternalTransportStats> { /*TODO*/ }
class NettyServerTransport implements InternalInstrumented<InternalTransportStats> { /*TODO*/ }

"core_internal": package io.grpc.internal
========================
public class ManagedChannelImpl implements InternalInstrumented<InternalChannelStats> {
  public ManagedChannelImpl(/* args */) {
    /* initialize it  */

    // in practice, we should not pass 'this' around before ctor is done
    InternalStatsServer.registerChannel(this);
  }
}
```